Exceptions in io.quarkiverse.operatorsdk.deployment.AddClusterRolesDecorator with QOSDK 6.8.0

I'm just updating QOSDK from 6.7.3 to 6.8.0 and noticed these exceptions in the `AddClusterRolesDecorator` during a `quarkusBuild` run.
```java
Ignoring com.hivemq.platform.operator.dependants.StatefulSetResource for generic resource role processing as it cannot be instantiated
io.javaoperatorsdk.operator.OperatorException: Couldn't instantiate KubernetesDependentResource 'com.hivemq.platform.operator.dependants.StatefulSetResource'. Context: AddClusterRolesDecorator
	at io.javaoperatorsdk.operator.api.config.Utils.instantiateAndConfigureIfNeeded(Utils.java:237)
	at io.javaoperatorsdk.operator.api.config.Utils.instantiate(Utils.java:255)
	at io.quarkiverse.operatorsdk.deployment.AddClusterRolesDecorator.lambda$getClusterRolePolicyRulesFromDependentResources$4(AddClusterRolesDecorator.java:134)
	at java.base@21.0.4/java.util.HashMap.forEach(HashMap.java:1429)
	at io.quarkiverse.operatorsdk.deployment.AddClusterRolesDecorator.getClusterRolePolicyRulesFromDependentResources(AddClusterRolesDecorator.java:109)
	at io.quarkiverse.operatorsdk.deployment.AddClusterRolesDecorator.createClusterRole(AddClusterRolesDecorator.java:76)
	at io.quarkiverse.operatorsdk.deployment.AddClusterRolesDecorator.lambda$visit$0(AddClusterRolesDecorator.java:58)
	at java.base@21.0.4/java.util.HashMap$Values.forEach(HashMap.java:1073)
	at io.quarkiverse.operatorsdk.deployment.AddClusterRolesDecorator.visit(AddClusterRolesDecorator.java:57)
	at io.quarkiverse.operatorsdk.deployment.AddClusterRolesDecorator.visit(AddClusterRolesDecorator.java:32)
	at io.fabric8.kubernetes.api.builder.Visitor.visit(Visitor.java:30)
	at io.fabric8.kubernetes.api.builder.VisitorWiretap.visit(VisitorWiretap.java:49)
	at io.fabric8.kubernetes.api.builder.Visitable.accept(Visitable.java:53)
	at io.fabric8.kubernetes.api.builder.Visitable.accept(Visitable.java:39)
	at io.fabric8.kubernetes.api.builder.Visitable.accept(Visitable.java:35)
	at io.dekorate.ResourceRegistry.lambda$generate$2(ResourceRegistry.java:181)
	at java.base@21.0.4/java.util.HashMap.forEach(HashMap.java:1429)
	at io.dekorate.ResourceRegistry.generate(ResourceRegistry.java:173)
	at io.dekorate.Session.generate(Session.java:293)
	at io.dekorate.Session.close(Session.java:256)
	at io.quarkus.kubernetes.deployment.KubernetesProcessor.lambda$build$5(KubernetesProcessor.java:203)
	at java.base@21.0.4/java.util.Optional.ifPresent(Optional.java:178)
	at io.quarkus.kubernetes.deployment.KubernetesProcessor.build(KubernetesProcessor.java:146)
	at java.base@21.0.4/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:854)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
	at java.base@21.0.4/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
Caused by: java.lang.IllegalStateException: Couldn't find a no-arg constructor for com.hivemq.platform.operator.dependants.StatefulSetResource
	at io.javaoperatorsdk.operator.api.config.Utils.getConstructor(Utils.java:247)
	at io.javaoperatorsdk.operator.api.config.Utils.instantiateAndConfigureIfNeeded(Utils.java:226)
	... 31 more
Caused by: java.lang.NoSuchMethodException: com.hivemq.platform.operator.dependants.StatefulSetResource.<init>()
Caused by: java.lang.IllegalStateException: Couldn't find a no-arg constructor for com.hivemq.platform.operator.dependants.StatefulSetResource
```

We get this for basically all of our dependent resources, because they use dependency injection in their constructors (which is supported to my knowledge).

For example, the resource from that stacktrace looks like this:
```java
@Singleton
@KubernetesDependent(labelSelector = HiveMQPlatformReconciler.LABEL_SELECTOR)
public class StatefulSetResource extends CRUDKubernetesDependentResource<StatefulSet, HiveMQPlatform> {

    private final @NotNull HiveMQPlatformOperatorProperties properties;
    private final @NotNull HiveMQPlatformReconcilerConfig config;

    StatefulSetResource(
            final @NotNull HiveMQPlatformOperatorProperties properties,
            final @NotNull HiveMQPlatformReconcilerConfig config) {
        super(StatefulSet.class);
        this.properties = properties;
        this.config = config;
    }
}
```

I would appreciate feedback if we're using DI in a wrong way or if this is a bug in the `AddClusterRolesDecorator` or `Utils` class. This doesn't happen with QOSDK 6.7.3 (just double checked).

Everything still seems to work though. The Quarkus build is successful and all of our tests are still green. The build window in IntelliJ just looks like the NCC-1701 on red alert.
![image](https://github.com/user-attachments/assets/2d39bc33-2a07-4332-9481-01f26b6bb61b)

--- a/core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/AddClusterRolesDecorator.java
+++ b/core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/AddClusterRolesDecorator.java
@@ -159,7 +159,7 @@ public class AddClusterRolesDecorator extends ResourceProvidingDecorator<Kuberne
                     } catch (Exception e) {
                         ignore = true;
                         log.warn(" Ignoring dependent " + dependentResourceClass.getName()
-                                + "because it couldn't be instantiated as it doesn't provide a no-arg constructor, preventing its group and plural to be determined.");
+                                + " because it couldn't be instantiated as it doesn't provide a no-arg constructor, preventing its group and plural to be determined.");
                     }
                 }
 

--- a/core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/AddClusterRolesDecorator.java
+++ b/core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/AddClusterRolesDecorator.java
@@ -25,6 +25,7 @@ import io.javaoperatorsdk.operator.processing.dependent.Creator;
 import io.javaoperatorsdk.operator.processing.dependent.Updater;
 import io.javaoperatorsdk.operator.processing.dependent.kubernetes.GenericKubernetesDependentResource;
 import io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependentResource;
+import io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependentResourceConfig;
 import io.quarkiverse.operatorsdk.annotations.RBACVerbs;
 import io.quarkiverse.operatorsdk.runtime.DependentResourceSpecMetadata;
 import io.quarkiverse.operatorsdk.runtime.QuarkusControllerConfiguration;
@@ -123,15 +124,30 @@ public class AddClusterRolesDecorator extends ResourceProvidingDecorator<Kuberne
                     verbs.add(RBACVerbs.DELETE);
                 }
                 final var isCreator = Creator.class.isAssignableFrom(dependentResourceClass);
+                boolean shouldDoubleCheckPatch = false;
                 if (isCreator) {
                     verbs.add(RBACVerbs.CREATE);
+
+                    // PATCH verb is also needed when using SSA to be able to add the finalizer when creating the resource
+                    // optimistically add PATCH method if resource should use SSA as defined by configuration
+                    // note that this isn't the only thing that is taken into account in that determination but should
+                    // a good first approximation, full determination is done if we can instantiate the dependent, see below
+                    final Object dependentResourceConfig = spec.getDependentResourceConfig();
+                    if (dependentResourceConfig instanceof KubernetesDependentResourceConfig<?> kubernetesDependentResourceConfig) {
+                        if (kubernetesDependentResourceConfig.useSSA().orElse(false)) {
+                            verbs.add(RBACVerbs.PATCH);
+                            shouldDoubleCheckPatch = true;
+                        }
+                    }
                 }
 
                 // Check if we're dealing with typeless Kubernetes resource or if we need to deal with SSA
+                boolean ignore = false;
+                KubernetesDependentResource<?, ?> kubeResource = null;
                 if (KubernetesDependentResource.class.isAssignableFrom(dependentResourceClass)) {
                     try {
-                        @SuppressWarnings("rawtypes")
-                        var kubeResource = Utils.instantiate(
+                        //noinspection rawtypes
+                        kubeResource = Utils.instantiate(
                                 (Class<? extends KubernetesDependentResource>) dependentResourceClass,
                                 KubernetesDependentResource.class, ADD_CLUSTER_ROLES_DECORATOR);
 
@@ -140,22 +156,35 @@ public class AddClusterRolesDecorator extends ResourceProvidingDecorator<Kuberne
                             resourceGroup = gvk.getGroup();
                             resourcePlural = gvk.getPluralOrDefault();
                         }
-
-                        // if we use SSA and the dependent resource class is not excluded from using SSA, we also need PATCH permissions for finalizer when the dependent resource is creatable
-                        if (isCreator && cri.getConfigurationService().shouldUseSSA(kubeResource)) {
-                            verbs.add(RBACVerbs.PATCH);
-                        }
                     } catch (Exception e) {
-                        log.warn("Ignoring " + dependentResourceClass.getName()
-                                + " for generic resource role processing as it cannot be instantiated", e);
+                        ignore = true;
+                        log.warn(" Ignoring dependent " + dependentResourceClass.getName()
+                                + "because it couldn't be instantiated as it doesn't provide a no-arg constructor, preventing its group and plural to be determined.");
                     }
                 }
-                final var dependentRule = new PolicyRuleBuilder()
-                        .addToApiGroups(resourceGroup)
-                        .addToResources(resourcePlural);
 
-                dependentRule.addToVerbs(verbs.toArray(String[]::new));
-                rules.add(dependentRule.build());
+                if (!ignore) {
+                    // if we need to double check if we really should use SSA
+                    if (shouldDoubleCheckPatch) {
+                        // we can only check if we managed to instantiate the dependent, though
+                        if (kubeResource != null) {
+                            if (!cri.getConfigurationService().shouldUseSSA(kubeResource)) {
+                                verbs.remove(RBACVerbs.PATCH);
+                            }
+                        } else {
+                            // if we couldn't double check, warn the user
+                            log.warn("Couldn't verify that dependent " + dependentResourceClass.getName()
+                                    + " really needs PATCH permission for SSA because it couldn't be instantiated. This means that a PATCH verb might have been added to the rule (group: "
+                                    + resourceGroup + " / plural: " + resourcePlural + ") when not needed.");
+                        }
+                    }
+
+                    final var dependentRule = new PolicyRuleBuilder()
+                            .addToApiGroups(resourceGroup)
+                            .addToResources(resourcePlural);
+                    dependentRule.addToVerbs(verbs.toArray(String[]::new));
+                    rules.add(dependentRule.build());
+                }
             }
         });
         return rules;

--- a/core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/AddClusterRolesDecorator.java
+++ b/core/deployment/src/main/java/io/quarkiverse/operatorsdk/deployment/AddClusterRolesDecorator.java
@@ -129,9 +129,11 @@ public class AddClusterRolesDecorator extends ResourceProvidingDecorator<Kuberne
                     verbs.add(RBACVerbs.CREATE);
 
                     // PATCH verb is also needed when using SSA to be able to add the finalizer when creating the resource
-                    // optimistically add PATCH method if resource should use SSA as defined by configuration
-                    // note that this isn't the only thing that is taken into account in that determination but should
-                    // a good first approximation, full determination is done if we can instantiate the dependent, see below
+                    // Here, we optimistically add PATCH method if the resource configuration states that SSA should be
+                    // used, despite this not being a correct/complete determination of whether the resource actually
+                    // uses SSA. This can only be determined by instantiating the dependent, which is why, if we can 
+                    // instantiate it, we double-check the SSA status later on and remove the PATCH method if we can
+                    // actually determine that it's not needed
                     final Object dependentResourceConfig = spec.getDependentResourceConfig();
                     if (dependentResourceConfig instanceof KubernetesDependentResourceConfig<?> kubernetesDependentResourceConfig) {
                         if (kubernetesDependentResourceConfig.useSSA().orElse(false)) {

